### DISCUSSION TITLE
Tiny Startup setup() after power on takes time?

### DISCUSSION THREAD

**Discussion by coderatchance** [SpenceKonde/megaTinyCore#1092] [Q&A]
Hello all, 

I have realised that my tiny takes like a second to startup, indeed, we can adjust that in the TinyMegaCore menu, 
however, is there anything else that is reducing the startup time? 
Actually, how much delay() do you actually use once you run a particular functions in setup().

For instance, I have to run the sensor setup with several adjustments, while I need to read from the eeprom as well... 
so what would you suggest is the minimum time needed to write to the register for the peripheral on the i2c...

Best.

**Comment by hmeijdam** [comment]
Are you using a bootloader?
Optiboot takes 1 second before it moves on to your sketch, if you change the UPDI pin to a reset pin..

![image](https://github.com/SpenceKonde/megaTinyCore/assets/9595465/ba4da665-5ab2-4849-a690-e1689a361a14)

**Reply by coderatchance** [reply]
Thank you ! I am not using the bootloader, 
went around the code, and got down to a reasonable delay, like 1/2 second,
however, I am still wondering, what are the reasonable delays when programming the peripherals...

**Reply by hmeijdam** [reply]
The peripherals in the tiny itself? The voltage reference may need a 25 microseconds to settle and with the ADC you best discard the first measurement. Other than that I don't think you need to delay anything, but to be sure consult the datasheet.

Waiting for external peripherals to be set up, totally depends on what you connect to your Attiny. I once had a pressure sensor that needed to warm up a bit before the readings were accurate. So I let it read pressure 100 times before using it. The longest startup delay I ever had to implement was one second, waiting for a Model aircraft receiver to go through it's gyro initialization.

**Reply by coderatchance** [reply]
The peripheral is suppose to be the the sensor on the i2c :) you are right, I am doing similar, measuring with the sensor and then obtaining the values and using further... hence I need like 100ms to measure... otherwise it does not even want to start...
Anyway, will remove other delays, I am really wondering, if it can setup everything else that quickly...

**Reply by hmeijdam** [reply]
At 20MHz clock you get about 2 Million CPU instructions done in 100ms. If that ain't enough to set up your program, I am curious what application you are building. Launching a spaceship?
